Document CoreCLR runtime provider events (#20817)
dotnet/docs: docs/fundamentals/diagnostics/runtime-thread-events.md

---
title: "ThreadPool runtime events"
description: See .NET runtime thread pool events that collect diagnostic information about thread pool in .NET Core. Thread pool events are worker thread pool events or I/O thread pool events.

> ms.date: "11/13/2020"

Reviewer: ```suggestion
ms.date: "11/13/2020"
```

ms.topic: reference
helpviewer_keywords
  - "ThreadPool events [.NET Core]"
  - "ETW, thread pool events (CoreCLR)"

> ms.assetid: f2a21e3a-3b6c-4433-97f3-47ff16855ecc

Reviewer: I think this needs to be deleted

```suggestion
```

|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CLR or CoreCLR.|

## ThreadPoolWorkerThreadStop event

|Keyword for raising the event|Level|
|-----------------------------------|-----------|-----------|
|`ThreadingKeyword` (0x10000)|Informational (4)|

|Event|Event ID|Description|
|-----------|--------------|-----------------|
|`ThreadPoolWorkerThreadStop`|51|A worker thread is stopped.|

|Field name|Data type|Description|
|----------------|---------------|-----------------|
|`ActiveWorkerThreadCount`|`win:UInt32`|Number of worker threads available to process work, including those that are already processing work.|
|`RetiredWorkerThreadCount`|`win:UInt32`|Number of worker threads that are not available to process work, but that are being held in reserve in case more threads are needed later.|
|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CLR or CoreCLR.|

## ThreadPoolWorkerThreadWait event

|Keyword for raising the event|Level|
|-----------------------------------|-----------|-----------|
|`ThreadingKeyword` (0x10000)|Informational (4)|

|Event|Event ID|Description|
|-----------|--------------|-----------------|

> |`ThreadPoolWorkerThreadStop`|51|A worker thread is waiting for work.|

Reviewer 2: ```suggestion
|`ThreadPoolWorkerThreadWait`|57|A worker thread starts waiting for work.|
```

|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CLR or CoreCLR.|

## ThreadPoolWorkerThreadRetirementStart event

|Keyword for raising the event|Level|
|-----------------------------------|-----------|-----------|
|`ThreadingKeyword` (0x10000)|Informational (4)|

|Event|Event ID|Description|
|-----------|--------------|-----------------|
|`ThreadPoolWorkerThreadRetirementStart`|52|A worker thread retires.|

|Field name|Data type|Description|
|----------------|---------------|-----------------|
|`ActiveWorkerThreadCount`|`win:UInt32`|Number of worker threads available to process work, including those that are already processing work.|
|`RetiredWorkerThreadCount`|`win:UInt32`|Number of worker threads that are not available to process work, but that are being held in reserve in case more threads are needed later.|
|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CLR or CoreCLR.|

## ThreadPoolWorkerThreadRetirementStop event

|Keyword for raising the event|Level|
|-----------------------------------|-----------|-----------|
|`ThreadingKeyword` (0x10000)|Informational (4)|

|Event|Event ID|Description|
|-----------|--------------|-----------------|

> |`ThreadPoolWorkerThreadStop`|53|A retired worker thread becomes active again.|

Reviewer 2: ```suggestion
|`ThreadPoolWorkerThreadRetirementStop`|53|A retired worker thread becomes active again.|
```

|`NewThreadWaveMagnitude`|`win:UInt16`|The magnitude of future variations in active thread count.|
|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CLR or CoreCLR.|

## ThreadPoolEnqueue event

 The following table shows the keyword and level.

|Keyword for raising the event|Level|
|-----------------------------------|-----------|
|`ThreadingKeyword` (0x10000)|Verbose (5)

 The following table shows the event information.

|Event|Event ID|Description|
|-----------|--------------|-----------------|
|`ThreadPoolEnqueue`|61|A work item was enqueued in the thread pool queue.|

 The following table shows the event data

|Field name|Data type|Description|
|----------------|---------------|-----------------|
|`WorkID`|`win:Pointer`|Pointer to the work request.|
|`ClrInstanceID`|`win:UInt16`|Unique ID for the instance of CoreCLR.|

## ThreadPoolDequeue event

 The following table shows the keyword and level.

|Keyword for raising the event|Level|
|-----------------------------------|-----------|
|`ThreadingKeyword` (0x10000)|Verbose (5)

 The following table shows the event information.

|Event|Event ID|Description|
|-----------|--------------|-----------------|

> |`ThreadPoolEnqueue`|61|A work item was dequeued from the thread pool queue.|

Reviewer 2: ```suggestion
|`ThreadPoolDequeue`|62|A work item was dequeued from the thread pool queue.|
```
